### PR TITLE
Add a buffer parameter to __init__

### PR DESCRIPTION
**Pull request by tgarc**
Allow usage of a preallocated buffer for the ringbuffer by adding a
``buffer`` parameter to ``__init__``

### CODE REVIEW

**Comment by mgeier** [comment]
That looks good, but shouldn't the `size` be calculated from the buffer's size?

I think it would be nice to also support CData pointers, which IIRC would raise an error in `from_buffer()` (see implementation of `write()` and `readinto()`).

**Comment by tgarc** [comment]
I agree it makes sense to add the try: except:. But I didn't want to mess with trying to estimate the size because it would've made the implementation more complicated. I could give it a shot though. You're talking about estimating the size only right? Or do you think it should attempt to guess the elementsize too?

**Comment by mgeier** [comment]
Yes, `size` only, otherwise it would be ambiguous.
I think it makes sense for the user to specify `elementsize`, while the `size` is calculated automatically from the length of the given buffer.

I don't think it's much more complicated, since you would have had to calculate this internally anyway to make sure that the buffer passed to `PaUtil_InitializeRingBuffer()` is big enough, otherwise you would have opened the door for segmentation faults.

**Comment by mgeier** [comment]
Thanks for the changes, that's great!
I'm just wondering if we should also raise an error if both `size` and `buffer` is given?

To reduce the amount of code we could probably do some fancy check like

```python
if (size, buffer).count(None) != 1:
    raise TypeError('exactly one of {size, buffer} is required')
```

... but probably that's too fancy?

BTW, I think it's a `TypeError` if a number is required and `NoneType` is given, isn't it?

**Comment by tgarc** [comment]
So if both size and buffer are passed I raise a ValueError which I think is the most appropriate. But otherwise I've added the changes you've suggested.

While I was doing this though I realized an issue that allowing a ``buffer`` argument introduces...what if the buffer passed is readonly? I did some tests and if you pass a ``bytes`` buffer (read-only), CFFI will still allow writing to the underlying C data. I don't think there's a concept of a ``const`` pointer in CFFI, in part because of support for PyPy. So this needs to be addressed in some manner...even if it's just a warning in the documentation. Note this is documented in the ``from_buffer`` documentation: http://cffi.readthedocs.io/en/release-1.4/using.html#ffi-interface

File: src/pa_ringbuffer.py [hunk 1]
```diff
@@ -73,15 +73,33 @@ class from the standard library can be used.
...
+    def __init__(self, elementsize, size=None, buffer=None):
         self._ptr = self._ffi.new('PaUtilRingBuffer*')
-        self._data = self._ffi.new('unsigned char[]', size * elementsize)
+        if buffer is None:
+            if size is None:
+                raise TypeError(
+                    "size is required when buffer parameter is not specified")
+            self._data = self._ffi.new('unsigned char[]', size * elementsize)
+        elif size is not None:
+            raise TypeError('exactly one of {size, buffer} is required')
```
Comment (mgeier) [inline src/pa_ringbuffer.py]: please remove trailing spaces

**Comment by mgeier** [comment]
Thanks a lot!

The thing with read-only buffers is really interesting. I tried using a `bytes` object as `buffer` and it seems to work, it actually becomes writable!
Since `from_buffer()` doesn't seem to give us any information about constness, I guess the only thing we can do is to mention this in the docs. Perhaps it's enough if we write something like "optional pre-allocated (and writable) buffer"?

**Comment by tgarc** [comment]
Yeah it is interesting, I read some discussion on the cffi forums that it was an issue to implement read only (because of pypy I believe). I will just use the same disclaimer that the cffi docs gives for from_buffer

**Comment by mgeier** [comment]
Perfect, thanks a lot!

**Comment by mgeier** [comment]
A recent discussion on the CFFI mailing list: https://groups.google.com/forum/#!topic/python-cffi/nJhDmBur_TM

**Comment by tgarc** [comment]
Haha! Good find. I hadn't thought of the side effects.

**Pull request merged**